### ISSUE TITLE
Token problem

### ISSUE THREAD

**Issue by stalkerj** [SebastianOsinski/HomebridgePluginSamsungAirConditioner#43]
root@homebridge:/var/lib/homebridge $ hom ebridge-samsung-ac-get-token 10.0.1.225
IP:
10.0.1.225
Error occured: C0075AFA087F0000:error: 028 0007E:Diffie-Hellman routines:DH check pa rams_ex:modulus too small:../deps/openss1 /openss1/crypto/dh/dh_check.c:40:

What i did wrong if i got such error? How can i fix it?

**Comment by stalkerj** [comment]
I’m on the latest plugin version

**Comment by SebastianOsinski** [comment]
Which version of Node.js are you using? I'm currently on 18.16 and it's working perfectly:
<img width="732" alt="image" src="https://github.com/SebastianOsinski/HomebridgePluginSamsungAirConditioner/assets/2934986/faf8e6d3-f40f-4425-b8f8-30a825f52c10">

**Comment by stalkerj** [comment]
I’m on 18.12.1 now

**Comment by stalkerj** [comment]
And I also using homebridge in docker of Synology)

**Comment by SebastianOsinski** [comment]
@stalkerj Could you help me test something? Could you install temporary version of the plugin by running `npm --prefix "/var/lib/homebridge" add  "https://github.com/SebastianOsinski/HomebridgePluginSamsungAirConditioner.git#dh_fix"` in your Homebridge Terminal? This does basically the same thing as `hb-service add <plugin-name>` but allows to install not released version of the plugin. I suspect that some AC models use weaker security than others and that might be the cause that it works for me but not for you.

After you install this prerelease version, please run `homebridge-samsung-ac-get-token <ip_address>` again to see if it works with the fixes I made. If not, then we'll iterate further

**Comment by stalkerj** [comment]
![image](https://github.com/SebastianOsinski/HomebridgePluginSamsungAirConditioner/assets/17808225/17cf5bfa-f062-479c-84fc-d1dd1e579519)
 Here is what i got in installation process

**Comment by SebastianOsinski** [comment]
Hm, ok. This is some docker-related problem with git. Let's ditch this way of testing that.
Could you just access `/var/lib/homebridge/node_modules/homebridge-plugin-samsung-air-conditioner/bin/ac-get-token.js` and replace `ciphers: 'DEFAULT:@SECLEVEL=0'` with `ciphers: 'ALL:@SECLEVEL=0'` in line 24? After you save it, try running `homebridge-samsung-ac-get-token <ip_address>`. This will help me pinpoint the issue a bit.

**Comment by stalkerj** [comment]
I changed the line. Save file. Restart homebridge. But have same error

![image](https://github.com/SebastianOsinski/HomebridgePluginSamsungAirConditioner/assets/17808225/df9442b9-42e7-41d8-b8f8-39049bfa3f86)

**Comment by stalkerj** [comment]
![image](https://github.com/SebastianOsinski/HomebridgePluginSamsungAirConditioner/assets/17808225/2bb4cf52-6430-4ed9-bce4-512daac202ce)

**Comment by SebastianOsinski** [comment]
Could you try replacing `'ALL:@SECLEVEL=0'` with `'HIGH:!DH:!aNULL'`?

**Comment by stalkerj** [comment]
> Could you try replacing `'ALL:@SECLEVEL=0'` with `'HIGH:!DH:!aNULL'`?

I made it and here what i got now:
![image](https://github.com/SebastianOsinski/HomebridgePluginSamsungAirConditioner/assets/17808225/788b9dd0-1ea5-4be1-b18e-965f98085ba0)

**Comment by SebastianOsinski** [comment]
I have the last idea, however I'm not even sure if it has any right to work. Revert back to z`ALL:@SECLEVEL=0` and add `minDHSize: 512` at the end:

```
const options = { 
        pfx: fs.readFileSync(pfxPath), 
        port: 2878, 
        host: ipAddress, 
        rejectUnauthorized: false,
        secureProtocol: 'TLSv1_method',
        ciphers: 'ALL:@SECLEVEL=0',
        minDHSize: 512
    }
```

You can try manipulating this number with other values. As far as I know, they have to be powers of 2, so 512, 256, 128 etc.

If that does not work, then I am out of ideas to be honest.

BTW. What's your AC model number? I wonder if anyone else has the same problem with that specific model.

**Comment by InyoEndo** [comment]
Yeah ive got the same problem and tried the above to no effect. Using mim-h02 wifi unit. This all used to work, now just does not which sucks :D

**Comment by SebastianOsinski** [comment]
@InyoEndo Could you try with options as follows:
```
    const options = { 
        port: 2878, 
        host: ipAddress, 
        rejectUnauthorized: false,
        secureProtocol: 'TLSv1_method',
        ciphers: 'HIGH:!DH:!aNULL'
    }
```

Unfortunately the certificate and SSL version our ACs use are basically obsolete and that's why there are so many problems with it. The fact that different devices seem to have different quirks does not help either

**Comment by InyoEndo** [comment]
@SebastianOsinski
Error occured: C077838D2D7F0000:error:0280007E:Diffie-Hellman routines:DH_check_params_ex:modulus too small:../deps/openssl/openssl/crypto/dh/dh_check.c:40:

**Comment by InyoEndo** [comment]
@SebastianOsinski 

Progresses further, but get cert auth failure now

Temp branch DH_FIX
HIGH:!DH:!aNUL + no pfx
IP:  192.168.1.9
Power on the device within the next 30 seconds
Error occured: Failed authentication

**Comment by SebastianOsinski** [comment]
The plug-in does not work because it's not fixed yet. If we develop fix for the token script, then it can be applied to the plug-in itself. 

So after your testing it seems that changing ciphers and removing  pfx line works? If so, then it looks like some AC models need the certificate and some are better without it.

I'll prepare new plug-in version soon with a new configuration option for disabling passing pfx and using those different ciphers

**Comment by SebastianOsinski** [comment]
I just released version 3.1.0 which adds option for skipping certificate. Refer to README to see what arguments/config to pass to enable this.

**Comment by stalkerj** [comment]
![image](https://github.com/SebastianOsinski/HomebridgePluginSamsungAirConditioner/assets/17808225/214febbc-b52b-4a84-90db-84840733c8ab)

![image](https://github.com/SebastianOsinski/HomebridgePluginSamsungAirConditioner/assets/17808225/3357d5e3-93fc-4a08-81d9-411c55a7e811)

**Comment by stalkerj** [comment]
Whats wrong now?

**Comment by SebastianOsinski** [comment]
The parameter should be `skip_certificate` not `skipCertificate`

**Comment by stalkerj** [comment]
its from your sample

![image](https://github.com/SebastianOsinski/HomebridgePluginSamsungAirConditioner/assets/17808225/83e806e6-b4ac-4f33-827c-8da747a4e796)

**Comment by SebastianOsinski** [comment]
I'll fix it then. Typo, should be `skip_certificate`

**Comment by stalkerj** [comment]
Great! It's working now.

My AC model - AR09HSSFRWKNER

**Comment by SebastianOsinski** [comment]
Great! Added to the list.

**Comment by SebastianOsinski** [comment]
Fixed by 3.1.0
